### ISSUE TITLE
0.17 release

### ISSUE THREAD

**Issue by gcp** [leela-zero/leela-zero#2275]
I think we should consider a 0.17 release with the Tensor Core support, batching, and the whole pile of bug-fixes and extensions to GTP support, as well as AutoGTP support for more experimentation.

Any reason why or why not? (If you have a massively important and useful pull you were about to request for example...)

If OK, I'll start working on it this week.

**Comment by l1t1** [comment]
it is better to release a beta version first. after all, many people dont compile the next branch. they just use the binary version.

**Comment by herazul** [comment]
Will it be possible with the release to have a little rundown (maybe in the readme ?) of what we know about how batching and tensor core impact performance, what to use and what corresponding options to use on what graphic cards, ...

**Comment by roy7** [comment]
/next basically is the beta test, but if we are able to get nightly build going for Windows binaries that would be awesome. @bood set something up to give us the OpenBLAS binaries needed that are newer than NuGet's modules as I understand it. So we'd just need LZ in AppVeyor configured to copy over his binaries and build the program and push the latest binaries to GitHub?

**Comment by gcp** [comment]
>what to use and what corresponding options to use on what graphic cards, ...

If the answer is not "don't do anything, use the defaults" then this is a serious bug that should be addressed first and foremost, not a burden to shift upon the user. 

I'm very strongly against shifting the burden of configuring for good performance to the user by exposing a bunch of settings (well, they're already exposed, I want to not even give the impression that this is something you have to touch) that they necessarily must tweak, with some risk they make the program weaker by not understanding how batching and threads influence search efficiency, etc. Note that documenting does nothing to stop that problem. The program should just work. And yes, you may get a tiny bit more speedup by finetuning provided you know *exactly* what you're doing, but there really shouldn't even be a temptation to do that. 

Currently if you launch with the defaults you'll get a nice speedup, something like 25% on mid-range cards, and a factor 2 or 3 with Tensor Cores.

And you don't need to do anything to get that.

**Comment by herazul** [comment]
Yeah i get that, makes a lot of sense. So nothing is needed in the readme then =)

With cards with tensor cores support, is there a big difference in performance between middle-end and high-end cards ? (2060 vs 2080ti for example ?)

**Comment by l1t1** [comment]
i suggest fix this thread at the top, and add a link of latest build link in the main post, such as https://ci.appveyor.com/project/zakki/ray/builds/21043153/artifacts

**Comment by gjm11** [comment]
Do I need to do anything to make it more likely that a fix for #2273 gets in? (E.g., further testing of the change described in there. My ability to do extensive testing is limited by my not having a very fast GPU.)

**Comment by l1t1** [comment]
pls spport more weights format ..
https://github.com/leela-zero/leela-zero/issues/2272

**Comment by Hersmunch** [comment]
Sounds good to me. It would be nice to get #2172 in so that any paused timed games are restarted as they left off but I'm not going to have time to get the unit test stuff working, at least for a little while. As long as next doesn't change too much then running any test can be done there whenever needed.

**Comment by gcp** [comment]
>With cards with tensor cores support, is there a big difference in performance between middle-end and high-end cards ? (2060 vs 2080ti for example ?)

The amount of Tensor Cores on those cards is scaled with the rest of the card resources, so the performance difference is what you would normally expect for all other applications too.

**Comment by gcp** [comment]
>pls spport more weights format ..

That doesn't look like it's close to getting a pull request so it's not even a consideration for 0.17.

**Comment by gcp** [comment]
>Do I need to do anything to make it more likely that a fix for #2273 gets in? 

Make a pull request so people can review it and protect the logic so it's not active when "dumbpass" is enabled?

**Comment by gjm11** [comment]
I think it should be active even when "dumbpass" is enabled. This isn't really about how smart we are about actually passing, it's about how we evaluate nodes. Without this fix, positions in which the NN wants to pass even though passing is stupid will be _misevaluated_ and this may lead to bad moves as well as unwise passes. (Further discussion should probably go in #2273 itself.)

**Comment by wonderingabout** [comment]
why not, if there are significant enough changes and fixes, no need to have major new features

**Comment by herazul** [comment]
@gcp The rate of promotion raised some question : Did you do a learning rate cut, or is the rate of promotion just because of -m30 or luck or something else ?

**Comment by gcp** [comment]
I didnt' change anything but I have wondered the same.

**Comment by gcp** [comment]
I'll leave some days to see if #2172 can get in (which just needs a reply/check), probably #2273 and a fix for #2280.

**Comment by zhanzhenzhen** [comment]
That's really great! Just a question: How much performance boost can we expect on a V100 card?

**Comment by Ishinoshita** [comment]
A benchmark on v0.16 vs v0.17 with different networks, that I just found on lifein19x19:

https://lifein19x19.com/viewtopic.php?f=18&t=16519&sid=46534765880903eefbf1b3208debfa4f

**Comment by l1t1** [comment]
where to download 0.17 binary  @Ishinoshita

**Comment by MarcelGruenauer** [comment]
@l1t1 I ran that benchmark; you have to compile it yourself. Speaking of which, this went well on macOS, but I wasn't able to compile it with Ubuntu on Google Cloud (P100).

"cmake .." says, among other things:

-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0

Then "cmake --build ." fails at the "Linking CXX executable tests" step with:

Scanning dependencies of target tests
[ 58%] Building CXX object CMakeFiles/tests.dir/src/tests/gtests.cpp.o
[ 60%] Building CXX object CMakeFiles/tests.dir/src/tests/utils_unittest.cpp.o
[ 62%] Linking CXX executable tests
gtest/googlemock/gtest/libgtest.a(gtest-all.cc.o): In function `testing::TestCase::TestCase(char const*, char const*, void (*)(), void (*)())':
gtest-all.cc:(.text+0x8d7d): undefined reference to `std::vector<int, std::allocator<int> >::~vector()'
gtest/googlemock/gtest/libgtest.a(gtest-all.cc.o): In function `testing::TestCase::~TestCase()':
gtest-all.cc:(.text+0x8e30): undefined reference to `std::vector<int, std::allocator<int> >::~vector()'
gtest-all.cc:(.text+0x8ea3): undefined reference to `std::vector<int, std::allocator<int> >::~vector()'
gtest/googlemock/gtest/libgtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::UnitTestImpl(testing::UnitTest*)':
gtest-all.cc:(.text+0x12827): undefined reference to `std::vector<int, std::allocator<int> >::~vector()'
gtest/googlemock/gtest/libgtest.a(gtest-all.cc.o): In function `testing::internal::UnitTestImpl::~UnitTestImpl()':
gtest-all.cc:(.text+0x129d6): undefined reference to `std::vector<int, std::allocator<int> >::~vector()'
gtest/googlemock/gtest/libgtest.a(gtest-all.cc.o):gtest-all.cc:(.text+0x12b09): more undefined references to `std::vector<int, std::allocator<int> >::~vector()' follow
collect2: error: ld returned 1 exit status
CMakeFiles/tests.dir/build.make:177: recipe for target 'tests' failed
make[2]: *** [tests] Error 1
CMakeFiles/Makefile2:69: recipe for target 'CMakeFiles/tests.dir/all' failed
make[1]: *** [CMakeFiles/tests.dir/all] Error 2
Makefile:149: recipe for target 'all' failed
make: *** [all] Error 2

Any ideas?

**Comment by gcp** [comment]
I'll test #2191 and then make a release candidate.

**Comment by MarcelGruenauer** [comment]
I tried various things - different compilers - and still can't get it to compile on Ubuntu on Google Cloud. Does anyone else have these issues?

**Comment by gcp** [comment]
>0.16 compiled fine a few months ago. Any ideas?

Please file a separate issue, it's offtopic here. FWIW it works fine on real Ubuntu.

**Comment by langit7** [comment]
> where to download 0.17 binary @Ishinoshita

http://fira.pw/lz/0.17/LZ_0.17-next-x64-28032019.zip

**Comment by l1t1** [comment]
my test1 next white vs master 7:3
```
# Black: Leela Zero
# BlackCommand: D:\tool\leela15\leelaz.exe --gtp -w D:\elfv2.gz --noponder -v 1601
# BlackLabel: Leela Zero:0.16
# BlackVersion: 0.16
# Date: March 28, 2019 4:37:37 PM CST
# Host: PC
# Komi: 7.5
# Referee: -
# Size: 19
# White: Leela Zero
# WhiteCommand: D:\LZ_0.17-next-x64-28032019\leelaz.exe --gtp -w D:\elfv2.gz --noponder -v 1601
# WhiteLabel: Leela Zero:0.16
# WhiteVersion: 0.16
# Xml: 0
#
#GAME	RES_B	RES_W	RES_R	ALT	DUP	LEN	TIME_B	TIME_W	CPU_B	CPU_W	ERR	ERR_MSG
0	B+R	B+R	B+R	0	-	91	919.2	949.1	0	0	0	
1	B+R	B+R	B+R	0	-	239	1587.7	1593.9	0	0	0	
2	W+R	W+R	W+R	0	-	98	401.7	363.8	0	0	0	
3	W+R	W+R	W+R	0	-	130	943.6	896.5	0	0	0	
4	W+R	W+R	W+R	0	-	290	1944.5	1940.7	0	0	0	
5	W+R	W+R	W+R	0	-	258	1900.4	1918.8	0	0	0	
6	W+R	W+R	W+R	0	-	146	1171.6	1229.8	0	0	0	
7	W+R	W+R	W+R	0	-	164	1243.5	1160.7	0	0	0	
8	B+R	B+R	B+R	0	-	151	1087	1102.9	0	0	0	
9	W+R	W+R	W+R	0	-	156	1249.7	1157.6	0	0	0	

```

**Comment by l1t1** [comment]
my test1 next black vs master 6:4
```
# Komi: 7.5
# Referee: -
# Size: 19
# White: Leela Zero
# WhiteCommand: D:\tool\leela15\leelaz.exe --gtp -w D:\elfv2.gz --noponder -v 1601
# WhiteLabel: Leela Zero:0.16
# WhiteVersion: 0.16
# Xml: 0
#
#GAME	RES_B	RES_W	RES_R	ALT	DUP	LEN	TIME_B	TIME_W	CPU_B	CPU_W	ERR	ERR_MSG
0	B+R	B+R	B+R	0	-	181	1441.7	1472.4	0	0	0	
1	B+R	B+R	B+R	0	-	175	1311.7	1351.9	0	0	0	
2	B+R	B+R	B+R	0	-	109	869.9	798.5	0	0	0	
3	B+R	B+R	B+R	0	-	157	762.5	749.7	0	0	0	
4	W+R	W+R	W+R	0	-	150	1121.9	1147.2	0	0	0	
5	B+R	B+R	B+R	0	-	157	1065.2	1202.7	0	0	0	
6	W+R	W+R	W+R	0	-	118	636.1	625.9	0	0	0	
7	W+R	W+R	W+R	0	-	148	1215.9	1207.6	0	0	0	
8	W+R	W+R	W+R	0	-	168	1465.8	1500	0	0	0	
9	B+R	B+R	B+R	0	-	155	1218.6	1247.9	0	0	0	
```

**Comment by Mbr18** [comment]
I have tested LZ_0.17-next-x64-28032019.zip on several different PC configurations and it works fine. 
On all configurations it shows: Using OpenCL batch size of 5 and Using 10 thread(s). Are these numbers fixed? One configuration is a Pentium G645. So I would expect 2 thread(s)?

**Comment by roy7** [comment]
@Mbr18 Batch size will default to half the number of threads, and I thought threads defaulted based on the # reported by your CPU.

**Comment by Mbr18** [comment]
On a q9400 configuration it also reports using 10 thread(s) where I would expect 4 threads. So Batch size 2 and 4 thread(s). V0.17 works fine so it is not a real problem.

**Comment by gcp** [comment]
I have everything pulled that I wanted, I will build the release candidates now.

**Comment by gcp** [comment]
Hmm, I found one annoying bug on RTX cards: if you launch autogtp, it will do a tuning run but not tune for fp16, so it won't actually use tensor cores. When it starts self-play, it will switch to tensor cores and retune (but appear to hang at first).

**Comment by gcp** [comment]
Suggested fix/workaround in #2312.

**Comment by gcp** [comment]
https://sjeng.org/dl/leela-zero-0.17-win64.zip

If there's no showstoppers, I'll tag and release this in a day or two.

**Comment by gcp** [comment]
...and here's another bug: AutoGTP always sends --batchsize, but CPU versions of the client won't recognize this.

I think I'll just make them ignore the option.

**Comment by gcp** [comment]
https://sjeng.org/dl/leela-zero-0.17-cpuonly-win64.zip

This includes #2313.

**Comment by gcp** [comment]
It seems NVIDIA's current driver packages no longer include 32-bit OpenCL support.

**Comment by roy7** [comment]
Someone from discord mentioned your exe has the debug DLLs linked (which they didn't have on their machine).

**Comment by OmnipotentEntity** [comment]
Specifically, you're linking against the debugging versions of VCRUNTIME14.dll (VCRUNTIME14D.dll) and ucrtbase.dll (ucrtbased.dll).  Which are not available via the Visual C++ Redistributable, and will probably not be available on most user's computers (as I believe they are only installed with Visual Studio itself).

**Comment by gcp** [comment]
Hmm, my build config most definitely doesn't have those. They must've been pulled in via another lib. I updated the OpenCL loader, so maybe that's the issue.

**Comment by gcp** [comment]
Yep, it's the new OpenCL DLL.

**Comment by gcp** [comment]
Packages are re-uploaded with a new DLL bundled. That problem should be fixed now.

**Comment by l1t1** [comment]
why @langit7's version has many files different from gcp?
```
D:\LZ_0.17-next-x64-28032019.rar
2015-04-07 15:57        331776        138451  libquadmath-0.dll
2019-03-28 14:45       1112576        388972  leelaz.exe
2015-04-07 15:57         82432         32230  libgcc_s_seh-1.dll
2015-04-07 15:57       1279488        284905  libgfortran-3.dll
2015-04-07 15:57      38340126       3583469  libopenblas.dll
```
gcp
```
D:\leela-zero-0.17-win64.zip
2019-04-02 19:11        235008         91579  autogtp.exe
2019-01-23 11:12        219596        124209  curl-ca-bundle.crt
2019-03-27 13:57       3582072       1755446  curl.exe
2017-08-31 20:17        102640         49601  gzip.exe
2019-04-03 01:34       3053568       1137033  leelaz.exe
2015-12-29 06:25        120334         50517  libgcc_s_dw2-1.dll
2018-10-22 19:31      20472300       4325027  libopenblas.dll
2015-12-29 06:25       1540622        438677  libstdc++-6.dll
2015-12-29 06:25         79360         33866  libwinpthread-1.dll
2017-10-05 21:34       2983837       1098720  msys-2.0.dll
2017-08-31 20:17        112382         48367  msys-gcc_s-1.dll
2019-04-03 05:10        102400         48910  OpenCL.dll
2018-08-18 07:56       6214656       2859032  Qt5Core.dll
2017-12-07 19:52           350           218  README.txt
2019-04-02 19:21        111104         44312  validation.exe
```

**Comment by iopq** [comment]
He didn't include autogtp, he just built leela-zero only

**Comment by l1t1** [comment]
he included libgcc_s_seh-1.dll, libquadmath-0.dll and libgfortran-3.dll,didnt include libgcc_s_dw2-1.dll

**Comment by l1t1** [comment]
seh and dw2 seems are two version of gcc,
https://sourceforge.net/projects/mingw-w64/files/Toolchains%20targetting%20Win64/Personal%20Builds/mingw-builds/8.1.0/ ,
but what the else two are

**Comment by gcp** [comment]
OpenBLAS can be built with some extra functionality that we don't require, and that can cause the compiler to pull in those extra DLL. I remove this in my build. You'll also note the OpenBLAS is 4 years more recent...

**Comment by l1t1** [comment]
suggest add support to vc++2019 complier

**Comment by gcp** [comment]
>suggest add support to vc++2019 complier

Feel free to submit pull requests, but this has got nothing to do with the release.

**Comment by gcp** [comment]
Release is out, closing issue.

**Comment by l1t1** [comment]
suggest to force update client, as it is much fast, so that  speed up the training process
